Added lint tests for notification sounds

TamasKiss opened the pull request:
Now running the tests will include checking every notification on actors.

lavjamanxd committed:
Map editor crashes - if items assigned to players without spawns

 - After a player removes a spawn point which is selected, the owner
   remains selected in the DropdownList after trying to spawn something
   the EditorActorBrush gets created with an invalid owner.
 - This fixes this by checking if the removed spawnpoint is the selected
   one, then falls back to the previous entry.

Commit:
Added lint tests for notification sounds

# Conflicts:
#	OpenRA.Game/Traits/LintAttributes.cs

GraionDilach commented on OpenRA.Mods.Common/Lint/CheckNotificationReferences.cs:
```diff
@@ -0,0 +1,65 @@
+			}
+		}
+
+		void CheckVoices(ActorInfo actorInfo, Action<string> emitError, Ruleset rules, string notification)
```
Please name this `CheckNotifications` instead.

GraionDilach commented:
Something is very suspicious to me about the notification check itself and Travis blew up the lint tests.

pchote commented on OpenRA.Mods.Common/Lint/CheckNotificationReferences.cs:
```diff
@@ -0,0 +1,65 @@
+
+		void CheckVoices(ActorInfo actorInfo, Action<string> emitError, Ruleset rules, string notification)
+		{
+			var soundInfo = rules.Notifications[notification.ToLowerInvariant()];
```
This needs to first check whether `notification` exists in `rules.Notifications`!

GraionDilach commented on OpenRA.Mods.Common/Lint/CheckNotifications.cs:
```diff
@@ -0,0 +1,68 @@
+			}
+		}
+
+		void CheckVoices(ActorInfo actorInfo, Action<string> emitError, Ruleset rules, string notification)
```
Again, please rename this to CheckNotifications. It's not voices per se what you're checking for.

GraionDilach commented on OpenRA.Mods.Common/Lint/CheckNotifications.cs:
```diff
@@ -0,0 +1,68 @@
+
+		void CheckVoices(ActorInfo actorInfo, Action<string> emitError, Ruleset rules, string notification)
+		{
+			if (!rules.Notifications.ContainsKey(notification))
```
Don't we want a lint error in this case as well? Asin "Actor's using undefined Notification string X!"?

GraionDilach commented:
Also, if you're updating a PR, please leave a comment on doing so, because GitHub does not notify people on new pushes, only on comments.

MustaphaTR commented on OpenRA.Mods.Common/Traits/Crates/CrateAction.cs:
```diff
@@ -27,6 +27,7 @@ public class CrateActionInfo : ITraitInfo
 		[PaletteReference] public readonly string Palette = "effect";
 
 		[Desc("Audio clip to play when the crate is collected.")]
+		[NotificationReferenceAttribute]
```
This is not actual Notification reference, but direct sound filename. Bot sure why it is caled Notification.

MustaphaTR commented:
Travis/make test gives errors for evwry map that has custom rules + 1 for main mod. So actual number of errors indeed wouldn't be that high.

TamasKiss commented:
CrateAction no longer contains the NotificationReferenceAttribute, as @MustaphaTR suggested. Thank you for the review.

I tried to poke around a bit since yesterday to see what is the root of the many errors, but could not find it. I will try some more tomorrow.

TamasKiss commented:
Finally got back to it, is `(!rules.Notifications.ContainsKey(notification))` really a cause for throwing an error? Imo a null there is normal.

pchote commented:
A null (i.e. no notification) should be allowed, but a reference to a notification type that isn't defined definitely shouldn't.

TamasKiss commented:
I updated the commit, not sure if I understood you right. 

btw the notifications that are causing these errors are mostly:
- UnitLost/Captured
- BuildingLost/Captured
But there are a few other offenders

pchote commented:
I see that a bunch of notifications that aren't tagged with the attribute yet.  I suggest using "Find References" `Game.Sound.PlayNotification` to list all the places that call it, and then from there find all the info fields that are passed in.

There's also an unintuitive style nit: the compiler contains some magic that supports omitting the `Attribute` part of the class name when it is used inside `[]`s, and that is the preferred c# code style.   So all the `[NotificationReferenceAttribute]`s should change to `[NotificationReference]` (but not the class name!).

pchote commented:
Any progress here @TamasKiss?

TamasKiss commented:
Sorry, but no change yet! I can do it in the afternoon though, it skipped my mind. 

Thanks for the reminder!

TamasKiss commented:
I added the extra notifications and renamed the usages. 

I also modified the namespace of RepairsUnits to mirror its directory.

Thanks for the help!

GraionDilach commented:
You need to rebase your branch.

TamasKiss commented:
It should be ready now

GraionDilach commented:
Think you mixed up branches.

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs:
```diff
@@ -94,6 +86,22 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
+
+				var orderedOwners = editorLayer.Players.Players.Values.ToList();
+				orderedOwners.Add(selectedOwner);
+				orderedOwners.Sort((x, y) => String.Compare(x.Name, y.Name, StringComparison.Ordinal));
```
> [SA1121] Use the built-in type alias 'string' rather than String or System.String

pchote commented:
I'll take this over once #15357 has been merged.

pchote commented:
Thanks for your help with setting up the base for this @TamasKiss!

I ended up reworking this a fair bit and didn't want to risk breaking your branches (you have other PR code in here) so I filed #15646 as a new PR instead of pushing it over this one.

Lets move the testing and discussion over to the new ticket.
